File: apps/windows/MyJobFinder.Core/WorkspaceLocator.cs
namespace MyJobFinder.Core;

public static class WorkspaceLocator
{
    public static string Find(string startPath)
    {
        var configured = Environment.GetEnvironmentVariable("MY_JOB_FINDER_WORKSPACE");
        if (!string.IsNullOrWhiteSpace(configured) && IsWorkspace(configured))
        {
            return Path.GetFullPath(configured);
        }

        var start = Path.GetFullPath(startPath);
        var bundledEngine = Path.Combine(start, "engine");
        if (IsWorkspace(bundledEngine)) return bundledEngine;

        var directory = new DirectoryInfo(start);
        while (directory is not null)
        {
            if (IsWorkspace(directory.FullName)) return directory.FullName;
            directory = directory.Parent;
        }

        throw new DirectoryNotFoundException(
            "The shared My Job Finder worker could not be located. Set MY_JOB_FINDER_WORKSPACE to the application workspace.");
    }

    private static bool IsWorkspace(string path)
    {
        return File.Exists(Path.Combine(path, "package.json"))
            && (File.Exists(Path.Combine(path, "src", "desktop", "cli.ts"))
                || File.Exists(Path.Combine(path, "dist", "desktop", "cli.js")));
    }
}

File: apps/macos/Sources/MyJobFinder/WorkspaceLocator.swift
import Foundation

enum WorkspaceLocator {
    static func find(startingAt startURL: URL = Bundle.main.bundleURL) throws -> URL {
        if let configured = ProcessInfo.processInfo.environment["MY_JOB_FINDER_WORKSPACE"] {
            let url = URL(fileURLWithPath: configured, isDirectory: true)
            if isWorkspace(url) { return url }
        }

        let bundled = startURL.appendingPathComponent("Contents/Resources/engine", isDirectory: true)
        if isWorkspace(bundled) { return bundled }
        let directBundle = startURL.appendingPathComponent("engine", isDirectory: true)
        if isWorkspace(directBundle) { return directBundle }

        var candidate = startURL.hasDirectoryPath ? startURL : startURL.deletingLastPathComponent()
        while candidate.path != "/" {
            if isWorkspace(candidate) { return candidate }
            candidate.deleteLastPathComponent()
        }
        throw CocoaError(.fileNoSuchFile, userInfo: [NSLocalizedDescriptionKey: "Set MY_JOB_FINDER_WORKSPACE to the folder containing the shared My Job Finder worker."])
    }

    private static func isWorkspace(_ url: URL) -> Bool {
        let package = url.appendingPathComponent("package.json").path
        let sourceWorker = url.appendingPathComponent("src/desktop/cli.ts").path
        let builtWorker = url.appendingPathComponent("dist/desktop/cli.js").path
        return FileManager.default.fileExists(atPath: package)
            && (FileManager.default.fileExists(atPath: sourceWorker) || FileManager.default.fileExists(atPath: builtWorker))
    }
}
